lib/mayaUsd/utils/trieVisitor.h

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright 2023 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef MAYA_USD_UTIL_TRIE_VISITOR_H
+#define MAYA_USD_UTIL_TRIE_VISITOR_H
+
+#include <mayaUsd/base/api.h>
+
+#include <ufe/sceneSegmentHandler.h>
+#include <ufe/trie.h>
+
+#include <functional>
+
+namespace MAYAUSD_NS_DEF {
+
+/// TrieVisitor allows visiting all nodes of a UFE Trie and receiving the full,
+/// correctly built UFE path of each node.
+///
+/// Note: this cannot be moved to UsdUfe since it needs to know about Maya run-time
+///       to build the UFE path segments with teh correct run-time ID.
+
+template <class T> struct TrieVisitor
+{
```

Comment:
Add this utility because composing the UFE paths from the UFE Trie nodes is surprisingly difficult and non-obvious, because they only contain UFE path component, not UFE path segments and thus do not contain the run-time ID. This is extremely awkward and unfortunate.

Reply:
This looks very USD-specific, which I guess that is okay, but maybe it would be better to rename this `UsdTrieVisitor`.  Because by construction our Ufe::Path's are 1 or 2 segments, as soon as you've hit a gateway node, you don't need to check for gateway nodes again: you know there won't be one.  Therefore, when traversing down the hierarchy, you could pass the path component separator and the run-time ID, with a `checkGateway = true` flag, then as soon as you find the gateway, change the path component separator to be the USD separator, switch the run-time ID to be USD, and change `checkGateway = false`.

Reply:
Okay for the naming.  The only concern I have is that repeatedly checking for gateway nodes could be a performance issue. We can always revisit that later if it turns out to be a perf problem.